src/concurrent_server.py

```diff
@@ -0,0 +1,162 @@
+    if request_list[0] == 'GET':
+        if request_list[2] == 'HTTP/1.1':
+            if 'Host: 127.0.0.1:' in split_request[1]:
+                return request_list[1]
+            else:
+                raise ValueError
+                # 400
+        else:
+            raise TypeError
+            # 505
+    else:
+        raise NameError
+        # 405
+
+
+def handle_listening(conn):
+    buffer_length = 4096
+    byte_msg = b''
+    decoded_msg = ""
+    while True:
+        part = conn.recv(buffer_length)
+        byte_msg += part
+        if len(part) < buffer_length:
+            decoded_msg = byte_msg.decode('utf-8')
+            break
+    return decoded_msg
+
+
+def response_ok(body, req_type):
+    response = response_template()
+    response[0] = response_check("200")
+    response[2] = u"Content-type: " + req_type + "; charset=utf-8\r\n"
+    response[4] = body
+    return response
+
+
+def response_err(req_type):
+    response = response_template()
+    response[0] = response_check(req_type)
+    response[4] = response_check(req_type)
+    return response
+
+
+def resolve_uri(uri, path='..'):
+    """
+    returns a body and type based on uri as a tuple
+    """
+    path_to_root = os.path.join(path, 'webroot', uri[1:])
+    print("path to root: ", path_to_root)
+    file_type = ""
+    if os.path.isfile(path_to_root):
+        print("is a file")
+        filepath = io.open(path_to_root, 'rb')
+        print("filepath :", filepath)
+        body = filepath.read()
+        print("body", body)
+        file_type = mimetypes.guess_type(uri)
+        print("file_type :", file_type[0])
+        filepath.close()
+        return body, file_type[0]
+    elif os.path.isdir(path_to_root):
+        print("is a directory", path_to_root)
+        return build_file_structre_html(path_to_root), file_type
+    else:
+        raise OSError
+
+
+def send_response(conn, response):
+    for c in response:
+        if isinstance(c, str):
+            conn.send(c.encode('utf-8'))
+        else:
+            conn.send(c)
+
+
+def server(conn, address):
+    try:
+        while True:
+            # listen on socket
+            client_request = handle_listening(conn)
+            print(client_request)
+            try:
+                uri = parse_request(client_request)
+                print('parsed uri: ', uri)
+            except ValueError:
+                client_response = response_err("400")
+            except TypeError:
+                client_response = response_err("505")
+            except NameError:
+                client_response = response_err("405")
+            try:
+                body, file_type = resolve_uri(uri)
+                print("body :", body)
+                print("file_type :", file_type)
+                client_response = response_ok(body, file_type)
+            except OSError:
+                client_response = response_err("404")
+            # Send the message
+            send_response(conn, client_response)
```

Comment:
I very much like the way that you've isolated the sending of a response to one line of code.  Good work.